Merge LoggingError into ErrorWrapper

`LoggingError` is the original error wrapper and is now OBE. Merge it into `ErrorWrapper`.

- [ ] Move `SyslogError` and `SetLoggerError` into `ErrorWrapper`.
- [ ] Add `new_...` functions.
- [ ] Add unit tests.
- [ ] Delete `LoggingError`.
- [ ] Fixup `avahi-alias`.
- [ ] Fixup `avahi-alias-daemon` (if needed).

Adding a enum variant to `ErrorWrapper` for `syslog::Error` causes significant dependency problems. Looks a conflict with `anyhow::Error`.

```
error[E0277]: `(dyn std::error::Error + Send + 'static)` cannot be shared between threads safely
   --> src/bin/avahi-alias-daemon.rs:39:74
    |
39  |             publish_aliases(&aliases_file, file_name, &new_modified_size)?;
    |                                                                          ^ `(dyn std::error::Error + Send + 'static)` cannot be shared between threads safely
    |
    = help: the trait `Sync` is not implemented for `(dyn std::error::Error + Send + 'static)`
    = note: required because of the requirements on the impl of `Sync` for `Unique<(dyn std::error::Error + Send + 'static)>`
    = note: required because it appears within the type `Box<(dyn std::error::Error + Send + 'static)>`
    = note: required because it appears within the type `Option<Box<(dyn std::error::Error + Send + 'static)>>`
    = note: required because it appears within the type `error_chain::State`
    = note: required because it appears within the type `syslog::errors::Error`
    = note: required because it appears within the type `ErrorWrapper`
    = note: required because of the requirements on the impl of `From<ErrorWrapper>` for `anyhow::Error`
    = note: required because of the requirements on the impl of `FromResidual<Result<Infallible, ErrorWrapper>>` for `Result<(), anyhow::Error>`
note: required by `from_residual`
   --> /Users/scott/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/try_trait.rs:339:5
    |
339 |     fn from_residual(residual: R) -> Self;
```

--- a/.gitignore
+++ b/.gitignore
@@ -10,6 +10,7 @@ Cargo.lock
 **/*.rs.bk
 
 # working directories
+data
 test-results
 
 # Temporary files

--- a/src/bin/avahi-alias-daemon.rs
+++ b/src/bin/avahi-alias-daemon.rs
@@ -8,7 +8,7 @@ use structopt::StructOpt;
 use anyhow::Result;
 use avahi_aliases::{
     init_console_logging, init_syslog_logging, AliasesFile, AvahiClient, AvahiRecord,
-    DaemonOpts, ErrorWrapper, LoggingError,
+    DaemonOpts, ErrorWrapper,
 };
 
 #[derive(PartialEq)]
@@ -19,7 +19,7 @@ struct ModifiedSize {
 
 #[paw::main]
 fn main(opts: DaemonOpts) {
-    match exec(opts) {
+    match inner_main(opts) {
         Ok(_) => std::process::exit(0),
         Err(error) => {
             log::error!("Error: {}", error);
@@ -28,7 +28,7 @@ fn main(opts: DaemonOpts) {
     }
 }
 
-fn exec(opts: DaemonOpts) -> Result<(), ErrorWrapper> {
+fn inner_main(opts: DaemonOpts) -> Result<(), ErrorWrapper> {
     init_logging(opts.common.verbose, opts.common.debug, opts.syslog)?;
     signon_app();
     let file_name = opts.common.file.as_str();
@@ -38,11 +38,11 @@ fn exec(opts: DaemonOpts) -> Result<(), ErrorWrapper> {
     Ok(())
 }
 
-fn init_logging(verbose: bool, debug: bool, syslog: bool) -> Result<(), LoggingError> {
+fn init_logging(verbose: bool, debug: bool, syslog: bool) -> Result<(), ErrorWrapper> {
     match syslog {
         true => init_syslog_logging(verbose, debug),
         false => {
-            init_console_logging(verbose, debug);
+            init_console_logging(verbose, debug)?;
             Ok(())
         },
     }

--- a/src/bin/avahi-alias.rs
+++ b/src/bin/avahi-alias.rs
@@ -8,17 +8,20 @@ use avahi_aliases::{
 
 #[paw::main]
 fn main(opts: CommandOpts) {
-    init_console_logging(opts.common.verbose, opts.common.debug);
-    let result = match opts.cmd {
-        Command::Add { aliases } => add(&opts.common.file, &aliases),
-        Command::List {} => list(&opts.common.file),
-        Command::Remove { aliases, force } => remove(&opts.common.file, &aliases, force),
-    };
-    if let Err(error) = result {
+    if let Err(error) = inner_main(opts) {
         log::error!("Error: {}", error);
         std::process::exit(1);
     }
     std::process::exit(0);
+}
+
+fn inner_main(opts: CommandOpts) -> Result<(), ErrorWrapper> {
+    init_console_logging(opts.common.verbose, opts.common.debug)?;
+    match opts.cmd {
+        Command::Add { aliases } => add(&opts.common.file, &aliases),
+        Command::List {} => list(&opts.common.file),
+        Command::Remove { aliases, force } => remove(&opts.common.file, &aliases, force),
+    }
 }
 
 fn add(filename: &str, arg_aliases: &[String]) -> Result<(), ErrorWrapper> {

--- a/src/error.rs
+++ b/src/error.rs
@@ -9,8 +9,6 @@ use std::io;
 
 use thiserror::Error;
 
-use crate::LoggingError;
-
 /// An enum to consolidate all errors passed to user-aware and logging code.
 #[derive(Error, Debug)]
 pub enum ErrorWrapper {
@@ -88,9 +86,14 @@ pub enum ErrorWrapper {
     },
 }
 
-/// Convert a `LoggingError` into an `ErrorWrapper`.
-impl From<LoggingError> for ErrorWrapper {
-    fn from(_: LoggingError) -> ErrorWrapper { ErrorWrapper::LoggingError }
+/// Convert a `SetLoggerError` into an `ErrorWrapper`.
+impl From<log::SetLoggerError> for ErrorWrapper {
+    fn from(_: log::SetLoggerError) -> ErrorWrapper { ErrorWrapper::LoggingError }
+}
+
+/// Convert a `SetLoggerError` into an `ErrorWrapper`.
+impl From<syslog::Error> for ErrorWrapper {
+    fn from(_: syslog::Error) -> ErrorWrapper { ErrorWrapper::LoggingError }
 }
 
 /// Convert a `dbus::Error` into an `ErrorWrapper`.
@@ -172,5 +175,304 @@ impl ErrorWrapper {
     where
         F: Into<String>, {
         ErrorWrapper::WriteError { file_name: file_name.into(), source }
+    }
+}
+
+//**********************************************************************************************
+// Unit tests
+//**********************************************************************************************
+
+#[cfg(test)]
+mod test {
+    use ErrorWrapper::*;
+
+    use super::*;
+
+    type R = Result<(), ErrorWrapper>;
+
+    static ALIAS: &str = "a0.local";
+    static ENUM_NAME: &str = "Enum";
+    static FILENAME: &str = "avahi-aliases";
+    static MESSAGE: &str = "message";
+
+    //******************************************************************************************
+    // ErrorWrapper::DBusError
+
+    #[test]
+    fn dbus_error_creation_works() {
+        if let Err(error) = Err(DBusError { message: MESSAGE.to_owned() }) as R {
+            let _ = format!("{}", error);
+        }
+    }
+
+    #[test]
+    fn from_dbus_error_creates_error_wrapper() {
+        let dbus_error = dbus::Error::new_failed(MESSAGE);
+        let _: ErrorWrapper = dbus_error.into();
+    }
+
+    #[test]
+    fn dbus_error_produces_correct_message() {
+        if let Err(error) = Err(DBusError { message: MESSAGE.to_owned() }) as R {
+            assert_eq!(format!("{}", error), format!(r#"D-Bus failure: {}"#, MESSAGE))
+        }
+    }
+
+    //******************************************************************************************
+    // ErrorWrapper::InvalidAliasError
+
+    #[test]
+    fn invalid_alias_error_creation_works() {
+        if let Err(error) = Err(InvalidAliasError { alias: ALIAS.to_owned() }) as R {
+            let _ = format!("{}", error);
+        }
+    }
+
+    #[test]
+    fn new_invalid_alias_error_creation_works() {
+        let _ = ErrorWrapper::new_invalid_alias_error(ALIAS);
+    }
+
+    #[test]
+    fn invalid_alias_error_produces_correct_message() {
+        if let Err(error) = Err(InvalidAliasError { alias: ALIAS.to_owned() }) as R {
+            assert_eq!(format!("{}", error), format!(r#"invalid alias: "{}""#, ALIAS))
+        }
+    }
+
+    //******************************************************************************************
+    // ErrorWrapper::InvalidAliasFileError
+
+    #[test]
+    fn invalid_alias_file_error_creation_works() {
+        if let Err(error) = Err(InvalidAliasFileError {
+            file_name: FILENAME.to_owned(),
+            alias: ALIAS.to_owned(),
+        }) as R
+        {
+            let _ = format!("{}", error);
+        }
+    }
+
+    #[test]
+    fn new_invalid_alias_file_error_creation_works() {
+        let _ = ErrorWrapper::new_invalid_alias_file_error(FILENAME, ALIAS);
+    }
+
+    #[test]
+    fn invalid_alias_file_error_produces_correct_message() {
+        if let Err(error) = Err(InvalidAliasFileError {
+            file_name: FILENAME.to_owned(),
+            alias: ALIAS.to_owned(),
+        }) as R
+        {
+            assert_eq!(
+                format!("{}", error),
+                format!(r#"invalid alias "{}" found in "{}""#, ALIAS, FILENAME)
+            )
+        }
+    }
+
+    //***************************************************************************************
+    // ErrorWrapper::LoggingError
+
+    #[test]
+    fn logging_error_creation_works() {
+        if let Err(error) = Err(LoggingError) as R {
+            let _ = format!("{}", error);
+        }
+    }
+
+    #[test]
+    fn logging_error_produces_correct_message() {
+        if let Err(error) = Err(LoggingError) as R {
+            assert_eq!(format!("{}", error), r#"logging error"#)
+        }
+    }
+
+    //***************************************************************************************
+    // ErrorWrapper::MetadataError
+
+    #[test]
+    fn metadata_error_creation_works() {
+        if let Err(error) = Err(MetadataError {
+            file_name: FILENAME.to_owned(),
+            source: io::Error::new(io::ErrorKind::AddrInUse, MESSAGE),
+        }) as R
+        {
+            let _ = format!("{}", error);
+        }
+    }
+
+    #[test]
+    fn new_metadata_error_creation_works() {
+        let _ = ErrorWrapper::new_metadata_error(
+            FILENAME,
+            io::Error::new(io::ErrorKind::AddrInUse, MESSAGE),
+        );
+    }
+
+    #[test]
+    fn metadata_error_produces_correct_message() {
+        if let Err(error) = Err(MetadataError {
+            file_name: FILENAME.to_owned(),
+            source: io::Error::new(io::ErrorKind::AddrInUse, MESSAGE),
+        }) as R
+        {
+            assert_eq!(
+                format!("{}", error),
+                format!(
+                    r#"could not get metadata for "{}": {}."#,
+                    FILENAME,
+                    io::Error::new(io::ErrorKind::AddrInUse, MESSAGE)
+                )
+            )
+        }
+    }
+
+    //***************************************************************************************
+    // ErrorWrapper::OpenError
+
+    #[test]
+    fn open_error_creation_works() {
+        if let Err(error) = Err(OpenError {
+            file_name: FILENAME.to_owned(),
+            source: io::Error::new(io::ErrorKind::AddrInUse, MESSAGE),
+        }) as R
+        {
+            let _ = format!("{}", error);
+        }
+    }
+
+    #[test]
+    fn new_open_error_creation_works() {
+        let _ = ErrorWrapper::new_open_error(
+            FILENAME,
+            io::Error::new(io::ErrorKind::AddrInUse, MESSAGE),
+        );
+    }
+
+    #[test]
+    fn open_error_produces_correct_message() {
+        if let Err(error) = Err(OpenError {
+            file_name: FILENAME.to_owned(),
+            source: io::Error::new(io::ErrorKind::AddrInUse, MESSAGE),
+        }) as R
+        {
+            assert_eq!(
+                format!("{}", error),
+                format!(
+                    r#"could not open "{}": {}."#,
+                    FILENAME,
+                    io::Error::new(io::ErrorKind::AddrInUse, MESSAGE)
+                )
+            )
+        }
+    }
+
+    //***************************************************************************************
+    // ErrorWrapper::EnumOutOfRangeError
+
+    #[test]
+    fn enum_out_of_range_error_creation_works() {
+        if let Err(error) =
+            Err(EnumOutOfRangeError { enum_name: ENUM_NAME.to_owned(), value: 0 }) as R
+        {
+            let _ = format!("{}", error);
+        }
+    }
+
+    #[test]
+    fn new_enum_out_of_range_error_creation_works() {
+        let _ = ErrorWrapper::new_enum_out_of_range_error(ENUM_NAME, 0);
+    }
+
+    #[test]
+    fn enum_out_of_range_error_produces_correct_message() {
+        if let Err(error) =
+            Err(EnumOutOfRangeError { enum_name: ENUM_NAME.to_owned(), value: 0 }) as R
+        {
+            assert_eq!(format!("{}", error), format!(r#"Invalid {} value: {}."#, ENUM_NAME, 0))
+        }
+    }
+
+    //***************************************************************************************
+    // ErrorWrapper::ReadError
+
+    #[test]
+    fn read_error_creation_works() {
+        if let Err(error) = Err(ReadError {
+            file_name: FILENAME.to_owned(),
+            source: io::Error::new(io::ErrorKind::AddrInUse, MESSAGE),
+        }) as R
+        {
+            let _ = format!("{}", error);
+        }
+    }
+
+    #[test]
+    fn new_read_error_creation_works() {
+        let _ = ErrorWrapper::new_read_error(
+            FILENAME,
+            io::Error::new(io::ErrorKind::AddrInUse, MESSAGE),
+        );
+    }
+
+    #[test]
+    fn read_error_produces_correct_message() {
+        if let Err(error) = Err(ReadError {
+            file_name: FILENAME.to_owned(),
+            source: io::Error::new(io::ErrorKind::AddrInUse, MESSAGE),
+        }) as R
+        {
+            assert_eq!(
+                format!("{}", error),
+                format!(
+                    r#"could not read "{}": {}."#,
+                    FILENAME,
+                    io::Error::new(io::ErrorKind::AddrInUse, MESSAGE)
+                )
+            )
+        }
+    }
+
+    //***************************************************************************************
+    // ErrorWrapper::WriteError
+
+    #[test]
+    fn write_error_creation_works() {
+        if let Err(error) = Err(WriteError {
+            file_name: FILENAME.to_owned(),
+            source: io::Error::new(io::ErrorKind::AddrInUse, MESSAGE),
+        }) as R
+        {
+            let _ = format!("{}", error);
+        }
+    }
+
+    #[test]
+    fn new_write_error_creation_works() {
+        let _ = ErrorWrapper::new_write_error(
+            FILENAME,
+            io::Error::new(io::ErrorKind::AddrInUse, MESSAGE),
+        );
+    }
+
+    #[test]
+    fn write_error_produces_correct_message() {
+        if let Err(error) = Err(WriteError {
+            file_name: FILENAME.to_owned(),
+            source: io::Error::new(io::ErrorKind::AddrInUse, MESSAGE),
+        }) as R
+        {
+            assert_eq!(
+                format!("{}", error),
+                format!(
+                    r#"could not write "{}": {}."#,
+                    FILENAME,
+                    io::Error::new(io::ErrorKind::AddrInUse, MESSAGE)
+                )
+            )
+        }
     }
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -13,7 +13,7 @@ pub use error::ErrorWrapper;
 mod line;
 pub use line::Line;
 mod logging;
-pub use logging::{init_console_logging, init_syslog_logging, LoggingError};
+pub use logging::{init_console_logging, init_syslog_logging};
 mod options;
 pub use options::{Command, CommandOpts, DaemonOpts};
 

--- a/src/logging.rs
+++ b/src/logging.rs
@@ -1,42 +1,30 @@
+//! Logging setup for `avahi-aliases`
+
 #![warn(clippy::all)]
 
-use thiserror::Error;
-
-/// LoggingError enumerates all errors returned by this module
-#[derive(Error, Debug)]
-pub enum LoggingError {
-    /// Represents a failure to connect to syslog.
-    #[error("Could not connect to syslog")]
-    SyslogError { source: syslog::Error },
-
-    /// Represents a failure setting the logger (syslog only)
-    #[error("Could not set logger")]
-    SetLoggerError { source: log::SetLoggerError },
-}
+use crate::ErrorWrapper;
 
 /// Initialize console logging
-pub fn init_console_logging(verbose: bool, debug: bool) {
+pub fn init_console_logging(verbose: bool, debug: bool) -> Result<(), ErrorWrapper> {
     env_logger::builder()
         .format_target(false)
         .format_level(false)
         .format_module_path(false)
         .format_timestamp(None)
         .filter(None, compute_log_level(verbose, debug))
-        .init();
+        .try_init()?;
+    Ok(())
 }
 
-pub fn init_syslog_logging(verbose: bool, debug: bool) -> Result<(), LoggingError> {
+pub fn init_syslog_logging(verbose: bool, debug: bool) -> Result<(), ErrorWrapper> {
     let formatter = syslog::Formatter3164 {
         facility: syslog::Facility::LOG_DAEMON,
         hostname: None,
         process: "avahi-alias-daemon".into(),
         pid: sysinfo::get_current_pid().unwrap(),
     };
-    let logger = syslog::BasicLogger::new(
-        syslog::unix(formatter).map_err(|source| LoggingError::SyslogError { source })?,
-    );
-    log::set_boxed_logger(Box::new(logger))
-        .map_err(|source| LoggingError::SetLoggerError { source })?;
+    let logger = syslog::BasicLogger::new(syslog::unix(formatter)?);
+    log::set_boxed_logger(Box::new(logger))?;
     log::set_max_level(compute_log_level(verbose, debug));
     Ok(())
 }
@@ -59,21 +47,39 @@ fn compute_log_level(verbose: bool, debug: bool) -> log::LevelFilter {
 
 mod tests {
 
+    use std::panic;
+
     use log;
+
+    use super::*;
 
     #[test]
     fn compute_log_level_yields_warn_as_default() {
-        assert_eq!(super::compute_log_level(false, false), log::LevelFilter::Warn);
+        assert_eq!(compute_log_level(false, false), log::LevelFilter::Warn);
     }
 
     #[test]
     fn compute_log_level_yields_info_for_verbose() {
-        assert_eq!(super::compute_log_level(true, false), log::LevelFilter::Info);
+        assert_eq!(compute_log_level(true, false), log::LevelFilter::Info);
     }
 
     #[test]
     fn compute_log_level_yields_debug_for_debug() {
-        assert_eq!(super::compute_log_level(false, true), log::LevelFilter::Debug);
-        assert_eq!(super::compute_log_level(true, true), log::LevelFilter::Debug);
+        assert_eq!(compute_log_level(false, true), log::LevelFilter::Debug);
+        assert_eq!(compute_log_level(true, true), log::LevelFilter::Debug);
+    }
+
+    #[test]
+    fn init_console_logging_works() {
+        init_console_logging(true, false).unwrap_or_else(|error| {
+            eprintln!(r#"Could completely test "init_console_logging": {:?}"#, error);
+        });
+    }
+
+    #[test]
+    fn init_syslog_logging_works() {
+        // There is a good chance that logging is already initialized. Catch the resulting
+        // panic. The result is less than perfect testing--c'est la guerre!
+        let _ = panic::catch_unwind(|| assert!(init_syslog_logging(true, false).is_ok()));
     }
 }

